### PR TITLE
Wrap cub in its own namespace

### PR DESCRIPTION
**Pull request by ngimel**
Tentative fix for #55027.
Wraps cub import in its name space so that static variables used by cub and thrust don't conflict if they end up in the different libraries when torch is built with BUILD_SPLIT_CUDA. cub variables end up in their own namespace, thrust variables are unwrapped, so they don't clash. 
This also allows extensions to use cub without wrapping it (thrust will still be problematic). The solution to allowing extensions to use thrust is to stop using thrust in pytorch completely. 
Now importing cub and importing thrust cannot coexist, so I had to move nonzero to its own file, and remove reliance on thrust functions for it. Nonzero now uses cub only. 
Also, we cannot selectively import just some of cub headers, we are forced to import `cub/cub.cuh`, which is not great. 
Caffe2 ops using cub are not touched (there are too many), so mixing caffe2 and torch will (can) still result in the same bug. We are moving towards disabling c2 ops, so I think this is fine. 
Still, even with that compiler (correctly) warns about redefinition of `CUB_NS_PREFIX` because including `ATen/ATen.h` transitively includes `thrust/complex.h` and that in turn includes original (empty) definition of `CUB_NS_PREFIX`. We probably can just ignore this warning. Here's an example warning:
```
In file included from /data/users/ngimel/pytorch/aten/src/ATen/native/cuda/Nonzero.cu:9:
/data/users/ngimel/pytorch/aten/src/ATen/cuda/CubUtils.cuh:4: warning: "CUB_NS_PREFIX" redefined
 #define CUB_NS_PREFIX namespace at{ namespace native{
 
In file included from /home/ngimel/local/cuda/include/thrust/system/cuda/config.h:76,
                 from /home/ngimel/local/cuda/include/thrust/system/cuda/detail/execution_policy.h:33,
                 from /home/ngimel/local/cuda/include/thrust/iterator/detail/device_system_tag.h:23,
                 from /home/ngimel/local/cuda/include/thrust/iterator/iterator_traits.h:111,
                 from /home/ngimel/local/cuda/include/thrust/detail/type_traits/pointer_traits.h:23,
                 from /home/ngimel/local/cuda/include/thrust/type_traits/is_contiguous_iterator.h:27,
                 from /home/ngimel/local/cuda/include/thrust/type_traits/is_trivially_relocatable.h:19,
                 from /home/ngimel/local/cuda/include/thrust/detail/complex/complex.inl:20,
                 from /home/ngimel/local/cuda/include/thrust/complex.h:1031,
                 from /data/users/ngimel/pytorch/c10/util/complex.h:9,
                 from /data/users/ngimel/pytorch/c10/core/ScalarType.h:4,
                 from /data/users/ngimel/pytorch/c10/core/Scalar.h:10,
                 from /data/users/ngimel/pytorch/build/aten/src/ATen/core/TensorBody.h:8,
                 from /data/users/ngimel/pytorch/aten/src/ATen/Tensor.h:3,
                 from /data/users/ngimel/pytorch/aten/src/ATen/Context.h:4,
                 from /data/users/ngimel/pytorch/aten/src/ATen/ATen.h:9,
                 from /data/users/ngimel/pytorch/aten/src/ATen/native/cuda/Nonzero.cu:1:
/home/ngimel/local/cuda/include/cub/util_namespace.cuh:43: note: this is the location of the previous definition
 #define CUB_NS_PREFIX
 
```
We will need a lint rule to prevent people from including `cub/cub.cuh`, because this will lead to #55027 reappearing again for some sequence of operations (and will lead to errors with cub code in extensions). 
Also, for this to work reliably we'll need to make sure that everything calling cub ends up in only one of libtorch_cuda_cu or libtorch_cuda_cpp, otherwise even namespace won't help (there still will be same symbols in 2 libraries). 

Upd: libtorch_cuda_cpp and libtorch_cuda_cu still contain the same symbols, which means that there exists a sequence of operations that will cause cache bug to reappear, so this is not a solution, we need to adjust file lists for BUILD_SPLITC_CUDA:
```
(pytorch) [ngimel@ ~/local/pytorch/build/lib] nm libtorch_cuda_cu.so | grep PerDeviceAttributeCache | c++filt
000000000c6bf808 u guard variable for at::native::cub::GetPerDeviceAttributeCache<at::native::cub::PtxVersionCacheTag>()::cache
000000000c600830 u guard variable for cub::GetPerDeviceAttributeCache<cub::PtxVersionCacheTag>()::cache
00000000018625e0 t at::native::cub::PerDeviceAttributeCache::DevicePayload at::native::cub::PerDeviceAttributeCache::operator()<at::native::cub::PtxVersion(int&)::{lambda(int&)#1}>(at::native::cub::PtxVersion(int&)::{lambda(int&)#1}&&, int)
00000000009ce630 t cub::PerDeviceAttributeCache::DevicePayload cub::PerDeviceAttributeCache::operator()<cub::PtxVersion(int&)::{lambda(int&)#1}>(cub::PtxVersion(int&)::{lambda(int&)#1}&&, int)
000000000c6bf820 u at::native::cub::GetPerDeviceAttributeCache<at::native::cub::PtxVersionCacheTag>()::cache
000000000c600840 u cub::GetPerDeviceAttributeCache<cub::PtxVersionCacheTag>()::cache
(pytorch) [ngimel@ ~/local/pytorch/build/lib] nm libtorch_cuda_cpp.so | grep PerDeviceAttributeCache | c++filt
0000000000ad2d98 u guard variable for at::native::cub::GetPerDeviceAttributeCache<at::native::cub::PtxVersionCacheTag>()::cache
0000000000ad2da0 u at::native::cub::GetPerDeviceAttributeCache<at::native::cub::PtxVersionCacheTag>()::cache
```
Upd2:
Moved TensorFactories.cu to torch_cuda_cu sources (see a change to caffe2/CMakeLists.txt), so now cub-related symbols are only in libtorch_cuda_cu. We'd need a test for that, any suggestions on how best to test it?
cc @zasdfgbnm @malfet

### CODE REVIEW

**Review by malfet — APPROVED** [review]

**Comment by ngimel** [comment]
@janeyx99 for windows linking error, we can either try to figure out why empty_cuda is failing (I have no idea about windows builds and exports, and will need help there), or we can pull randperm_cuda out of tensor factories and into its own file, and add that to libtorch_cuda_cu. The important thing is that randperm_cuda has to be in libtorch_cuda_cu. Which do you prefer?

**Comment by ngimel** [comment]
Remaining questions for this diff
1) try to fix windows linking error or pull out randperm_cuda out of TensorFactories.cu into libtorch_cuda_cu? (I'm inclined towards the latter) - @janeyx99, @malfet
2) how to create a lint rule and test to forbid `#include <cub/something.cuh>" - @samestep I guess?
3) how to create a test that makes sure that
`nm libtorch_cuda_cu.so | grep cache | c++filt | grep -w cub | grep -w cache` comes up empty for libtorch_cuda_cpp? @malfet
4) is `at::native` ok namespace for cub calls, or is it better to have something that has less probability of clashing with extensions (e.g. `at::native::pytorch_cub`?) That would require minor changes to cub callsites. 

Follow-up (I'll file a separate issue after this PR is landed):
get rid of thrust calls that result in non-namespaced cache variables, currently with this PR there are namespaced and non-namespaced versions:
```
000000000c796e80 u guard variable for at::native::cub::DeviceCount()::cache
000000000c796868 u guard variable for at::native::cub::GetPerDeviceAttributeCache<at::native::cub::PtxVersionCacheTag>()::cache
000000000c6d7ea0 u guard variable for cub::DeviceCount()::cache
000000000c6d7890 u guard variable for cub::GetPerDeviceAttributeCache<cub::PtxVersionCacheTag>()::cache
000000000c796e88 b at::native::cub::DeviceCount()::cache
000000000c796880 u at::native::cub::GetPerDeviceAttributeCache<at::native::cub::PtxVersionCacheTag>()::cache
000000000c6d7ea8 b cub::DeviceCount()::cache
000000000c6d78a0 u cub::GetPerDeviceAttributeCache<cub::PtxVersionCacheTag>()::cache
```

**Comment by janeyx99** [comment]
> @janeyx99 for windows linking error, we can either try to figure out why empty_cuda is failing (I have no idea about windows builds and exports, and will need help there), or we can pull randperm_cuda out of tensor factories and into its own file, and add that to libtorch_cuda_cu. The important thing is that randperm_cuda has to be in libtorch_cuda_cu. Which do you prefer?

The latter is fine!

**Comment by janeyx99** [comment]
I think what's important to us right now is that cudnn only depends on torch_cuda_cpp, as its static library is pretty big.

**Comment by janeyx99** [comment]
Also, for the lint rule, one can add a check similar to https://github.com/pytorch/pytorch/blob/master/.github/workflows/lint.yml#L44-L55 I believe?

**Comment by ngimel** [comment]
> Also, for the lint rule, one can add a check similar to https://github.com/pytorch/pytorch/blob/master/.github/workflows/lint.yml#L44-L55 I believe?

Yep, just did that https://github.com/pytorch/pytorch/pull/55292/commits/41b33f6e30ba7550cf29617f2bcbcc58198be9f2

File: .github/workflows/lint.yml [hunk 1]
```diff
@@ -67,6 +67,9 @@ jobs:
         run: |
           set -eux
           python torch/testing/check_kernel_launches.py |& tee ${GITHUB_WORKSPACE}/cuda_kernel_launch_checks.txt
+      - name: Ensure no direct cub include
+        run: |
+          (! git grep -I -l $'#include <cub/' --  ./aten  ':(exclude)aten/src/ATen/cuda/CubUtils.cuh' || (echo "The above files have direct cub include; please include ATen/cuda/CubUtils.cuh instead and wrap your cub calls in at::native namespace if necessary"; false))
```
Comment (janeyx99) [inline .github/workflows/lint.yml:72]: Is there a reason to prefer -l vs -no? It looks like -no shows more information, which might be better.

**Comment by lpasselin** [comment]
Is there any chance the fix will be backported to the LTS version?

**Pull request closed without merging**